src/graph/serializers/decoders/prev/v5/decode_graph.c

```diff
@@ -0,0 +1,166 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include <assert.h>
+#include "decode_v5.h"
+
+typedef enum {
+	V5_T_NULL = 0,
+	V5_T_STRING = 0x001,
+	V5_T_INT64 = 0x004,
+	V5_T_BOOL = 0x010, // shares 'longval' representation in SIValue union
+	V5_T_DOUBLE = 0x040,
+	V5_T_PTR = 0x080,
+	V5_T_CONSTSTRING = 0x100, // only used in deserialization routine
+	V5_T_NODE = 0x200,
+	V5_T_EDGE = 0x400,
+} PrevSIType;
+
+static SIType _ConvertSIType(PrevSIType prev_type) {
+	switch(prev_type) {
+	case V5_T_INT64:
+		return T_INT64;
+	case V5_T_DOUBLE:
+		return T_DOUBLE;
+	case V5_T_STRING:
+	case V5_T_CONSTSTRING:
+		return T_STRING;
+	case V5_T_BOOL:
+		return T_BOOL;
+	case V5_T_NULL:
+		return T_NULL;
+	default: // should not occur
+		assert(false);
+	}
+}
+
+static SIValue _RdbLoadSIValue_v5(RedisModuleIO *rdb) {
+	/* Format:
+	 * SIType
+	 * Value */
+	PrevSIType prev_type = RedisModule_LoadUnsigned(rdb);
+	SIType t = _ConvertSIType(prev_type); // Convert SIType to latest enum value
+
+	switch(t) {
+	case T_INT64:
+		return SI_LongVal(RedisModule_LoadSigned(rdb));
+	case T_DOUBLE:
+		return SI_DoubleVal(RedisModule_LoadDouble(rdb));
+	case T_STRING:
+		// Transfer ownership of the heap-allocated string to the
+		// newly-created SIValue
+		return SI_TransferStringVal(RedisModule_LoadStringBuffer(rdb, NULL));
+	case T_BOOL:
+		return SI_BoolVal(RedisModule_LoadSigned(rdb));
+	case T_NULL:
+	default: // currently impossible
+		return SI_NullVal();
+	}
+}
+
+static void _RdbLoadEntity_v5(RedisModuleIO *rdb, GraphContext *gc, GraphEntity *e) {
+	/* Format:
+	 * #properties N
+	 * (name, value type, value) X N
+	*/
+	uint64_t propCount = RedisModule_LoadUnsigned(rdb);
+	if(!propCount) return;
+
+	for(int i = 0; i < propCount; i++) {
+		char *attr_name = RedisModule_LoadStringBuffer(rdb, NULL);
+		SIValue attr_value = _RdbLoadSIValue_v5(rdb);
+		Attribute_ID attr_id = GraphContext_GetAttributeID(gc, attr_name);
+		assert(attr_id != ATTRIBUTE_NOTFOUND);
+		GraphEntity_AddProperty(e, attr_id, attr_value);
+		RedisModule_Free(attr_name);
+	}
+}
+
+static void _RdbLoadNodes_v5(RedisModuleIO *rdb, GraphContext *gc) {
+	/* Format:
+	 * #nodes
+	 *      #labels M
+	 *      (labels) X M
+	 *      #properties N
+	 *      (name, value type, value) X N
+	*/
+
+	uint64_t nodeCount = RedisModule_LoadUnsigned(rdb);
+	if(nodeCount == 0) return;
+
+	Graph_AllocateNodes(gc->g, nodeCount);
+	for(uint64_t i = 0; i < nodeCount; i++) {
+		Node n;
+
+		// Extend this logic when multi-label support is added.
+		// #labels M
+		uint64_t nodeLabelCount = RedisModule_LoadUnsigned(rdb);
+
+		// * (labels) x M
+		// M will currently always be 0 or 1
+		uint64_t l = (nodeLabelCount) ? RedisModule_LoadUnsigned(rdb) : GRAPH_NO_LABEL;
```

Comment:
I think we should plan a head and work with a for loop here, this will enable nodes with multiple labels.

Reply:
The RDB encoding is future-proofed, in that we store label count followed by label IDs 0...N.

I don't really see the benefit of looping here (or in the other `decode_graph` files) right now, as `Graph_CreateNode` only accepts one label ID. When we introduce multi-label support, I expect that signature will change, and the up-to-date serialization routines will need to be updated, but not the encoding version.